frontend/src/app/App.test.tsx

```diff
@@ -917,6 +917,51 @@ describe("App.onHistoryChange", () => {
       expect(instance.onPageChange).toHaveBeenLastCalledWith("sub_hash")
     })
   })
+
+  it("doesn't rerun when we are on the same page and the url contains an anchor", () => {
+    const pushStateSpy = jest.spyOn(window.history, "pushState")
+
+    const instance = wrapper.instance() as App
+    window.history.pushState({}, "", "#foo_bar")
+    jest.spyOn(instance, "onPageChange")
+    instance.onHistoryChange()
+
+    // check that onPageChange is not run when anchor is in the url and we are on the same page
+    expect(instance.onPageChange).not.toHaveBeenCalled()
+    // @ts-expect-error
+    window.history.pushState.mockClear()
+    pushStateSpy.mockRestore()
+  })
+
+  it("does rerun when we are navigating to a different page and the url contains an anchor", () => {
+    const pushStateSpy = jest.spyOn(window.history, "pushState")
+
+    const instance = wrapper.instance() as App
+    jest.spyOn(instance, "onPageChange")
+
+    // navigate to current page with anchor
+    window.history.pushState({}, "", "#foo_bar")
+    instance.onHistoryChange()
+    // check that onPageChange is not run when anchor is in the url and we are on the same page
+    expect(instance.onPageChange).not.toHaveBeenCalled()
+
+    // navigate to current page with anchor
+    window.history.pushState({}, "", "#foo_baz")
+    instance.onHistoryChange()
+    // check that onPageChange is not run when anchor is in the url and we are on the same page
+    expect(instance.onPageChange).not.toHaveBeenCalled()
+
+    instance.handleNewSession(
+      new NewSession({ ...NEW_SESSION_JSON, pageScriptHash: "page_2_hash" })
+    )
+    window.history.back()
+    instance.onHistoryChange()
+
+    expect(instance.onPageChange).toHaveBeenCalled()
```

Comment:
This should probably follow the pattern used in tests above where we don't manually call `instance.onHistoryChange()` (since it'll get called asynchronously as a result of calling `window.history.back()`) and use `waitFor` to watch for `onPageChange` getting called.

Also, we should be more specific with the exact script hash we expect `onPageChange` to be called with